src/components/Viewer.jsx

```diff
@@ -39,11 +57,148 @@ export class Viewer extends Component {
 
     this.gesturesHandler = new Hammer(this.viewer)
     this.gesturesHandler.on('swipe', this.onSwipe.bind(this))
+
+    this.gesturesHandler.get('pinch').set({ enable: true })
+    this.gesturesHandler.get('pan').set({ direction: Hammer.DIRECTION_ALL })
+
+    // During a gesture, everything is computed with a base value (the state of the image when the gesture starts) and a delta (a translation / zoom, described by the gesture). When a gesture starts, we record the current state of the image.
+    this.gesturesHandler.on('panstart', this.prepareForGesture.bind(this))
+    this.gesturesHandler.on('pinchstart', this.prepareForGesture.bind(this))
+
+    // during a pan, we add the gestures delta to the initial offset to get the new offset. The new offset is then scaled : if the pan distance was 100px, but the image was scaled 2x, the actual offset should only be 50px. FInally, this value is clamped to make sure the user can't pan further than the edges.
+    this.gesturesHandler.on('pan', e => {
```

Comment:
The more I see comments that try to express the code behavior, the more I think it should be abstract into a library.